### ISSUE TITLE
[Feature Request] Individual page duration

### ISSUE THREAD

**Issue by erkilic** [edward-shen/MMM-pages#46]
As an idea - for some users it might be helpful to adjust individuall duration time for each page.
This would have the advantage, that pages with more data can rest longer, where other pages with less information can be slided faster.
Furthermore the user would have the possibility to give a page the duration "0" seconds, if he doesn't wants to show the module although it should be loaded and operationg in the background (e.g. MPlayerRadio).
Would this new feature be possible? I would appreciate it very :-)

**Comment by edward-shen** [comment]
Sorry for the late reply, things have gotten a little hectic.

> it might be helpful to adjust individuall duration time for each page.

This seems like a good idea, but I'm a little busy right now to implement it. 

>  Furthermore the user would have the possibility to give a page the duration "0" seconds, if he doesn't wants to show the module although it should be loaded and operationg in the background (e.g. MPlayerRadio).

I believe this can be achieved either through not setting a position in the config for that module, or a combination where you set the module as a fixed module and hiding it through CSS.

**Comment by erkilic** [comment]
ok thx edward. 


⁣Mit freundlichen Grüßen
Gürkan Erkilic​

**Comment by soubhik-khan** [comment]
+1 to individual page duration

**Comment by edward-shen** [comment]
I'm currently looking for a new maintainer for my magic mirror modules, so I won't be implementing any new features myself. PRs are always appreciated, but don't expect feature requests to be implemented unless a new maintainer is found.

**Comment by sdetweil** [comment]
Added support in PR #73
same function logic in https://github.com/shbatm/MMM-Carousel and https://github.com/mezderman/MMM-ModulesGroupsRotation

**Comment by KristjanESPERANTO** [comment]
I also created a PR (#88). In my opinion, my approach (with a simple array) has two advantages: no major code changes were necessary and the configuration is somewhat clearer and therefore simpler.

Here for comparison what the configuration of my PR and the other PR look like (if I have understood the other PR correctly):

My PR #88:

```js
            individualRotationTimes: [
            	5000,	// 5 seconds for page 1
            	10000,	// 10 seconds for page 2
            	5000,	// 5 seconds for page 3
            	15000,	// 15 seconds for page 4
            ],
```

PR #76:

```js
            pageTimeout: [
            	{ pageNumber: 1, timeout: 5000 },	// 5 seconds for page 1
            	{ pageNumber: 2, timeout: 10000 },	// 10 seconds for page 2
            	{ pageNumber: 3, timeout: 5000 },	// 5 seconds for page 3
            	{ pageNumber: 4, timeout: 15000 },	// 15 seconds for page 4
            ],
```

**Comment by sdetweil** [comment]
The point of my page Number field was to not require the exact symmetry between extra info and specific page timing

Have 8 pages, want page 6 longer, takes 1 entry, not 8, and if you wanted to change another, add one entry, order not affecting the results.

**Comment by KristjanESPERANTO** [comment]
I can see the advantages of your approach. But I think the simplicity of my approach, both in the config and in the code, might outweigh it.

To illustrate, here are the configs of both approaches, where only the 8th page is different:

#### PR 76

```js
            pageTimeout: [
            	{ pageNumber: 8, timeout: 60000 },
            ],
```

#### PR 88

Long version:

```js
            individualRotationTimes: [
            	15000,	// page 1
            	15000,	// page 2
            	15000,	// page 3
            	15000,	// page 4
            	15000,	// page 5
            	15000,	// page 6
            	15000,	// page 7
            	60000,	// page 8
            ],
```

Short version:

```js
            individualRotationTimes: [ 15000, 15000,  15000, 15000, 15000, 15000, 15000, 60000 ],
```

<hr>

Without causing a breaking change or significant increase in code complexity, I can now only see these two practicable approaches. Even though I like the simplicity of my approach, I would also be okay with sdetweil's. @edward-shen What do you think?

**Comment by sdetweil** [comment]
You could also use 0 for the pages  that use the default timing mechanism

This new property is optional 

I did the same design implementation  for MMM-Carousel, and MMM-ModulesGroupsRotation
(carousel first)(trying to be consistent at the user experience)

**Comment by KristjanESPERANTO** [comment]
> You could also use 0 for the pages that use the default timing mechanism

Sure, but I can't see much advantage in that. I rather see that config and code would become more complex and error-prone.

> I did the same design implementation for MMM-Carousel ... (trying to be consistent at the user experience)

I also contribute to the "official" fork of [MMM-Carousel](https://github.com/shbatm/MMM-Carousel). I can't see this function there yet. If you have only done this in your fork, I consider the consistent argument to be invalid.

**Comment by edward-shen** [comment]
For bigger number of pages, I do like Sam's implementation, but I like how terse a list is. I am a fan of explicit over implicit. How about a mix?

```js
let timings = {
  "default": 3000,
  "1": 500,
  "4": 600,
};

timings[1] || timings["default"] // 500
timings[40] || timings["default"] // 3000
```

At the very least I would like to see an implementation where a default value is supported if the timing isn't found. Currently, neither implementation supports falling back.

Also bonus points if you add support for hidden/named pages.

**Comment by sdetweil** [comment]
Mine only uses the extra if specified, default otherwise. User only adds difference from default config

I thought named/hidden were by notification only. And not timed

**Comment by sdetweil** [comment]
@KristjanESPERANTO  Carousel was not updated for a long while too. so I made changes to my fork. I don't keep track of who is doing what to what.

**Comment by KristjanESPERANTO** [comment]
> ... How about a mix?
> 
> let timings = {
>   "default": 3000,
>   "1": 500,
>   "4": 600,
> };

@edward-shen I have changed my PR (#88) in this way 🙂 

> Also bonus points if you add support for hidden/named pages.

I don't completely understand this, can you explain?

**Comment by edward-shen** [comment]
MMM-pages have support for "named" pages in their config, e.g. as an arg to "hiddenPages". Unlike typical pages, each page has a string name instead of the page number. People can then send a request to change to these pages.

These hidden pages are not part of the typical rotation, but might be useful for, say, bringing up an "admin" panel on certain notifications.

I'm not actually sure if your PR supports hidden pages currently, but I think the way its written it won't affect that behavior at all.

**Comment by sdetweil** [comment]
but named/hidden were never in the rotation. are you proposing some timeout to return to rotation?

ive never seen this requested

**Comment by edward-shen** [comment]
Yes, there are currently no requests for hidden pages timeouts, but with more a flexible timings object, it could make sense to implement that as an extension.

Whether Kristjan wants to implement them is up to them :)

**Comment by edward-shen** [comment]
A separate pr is fine imo

**Comment by KristjanESPERANTO** [comment]
I just merged PR #88 and released a new version 🙂 That settles this Feature Request and I am closing it.

As I can't find the time to do it straight away, I opened issue #90 to pursue the idea with the timeout for hidden pages.
